### PR TITLE
Add documentation for tukit.conf (bsc#1192307)

### PR DESCRIPTION
**Pull request by aplanas**

### CODE REVIEW

**Comment by laenion** [comment]
I'll merge this for transactional-update 4.0 along with a man page for `tukit` itself (which is also still undocumented yet).

Just one note: I think we *never* want to add _/var/lib_ to the list of bind mounts.
The whole concept of transactional updates is based around the concept that an update can *not* break the currently running system. By bind-mounting such a vital directory that whole concept is basically nullified - and on the contrary will cause significant side effects:

/var/lib contains a lot of data, including databases, docker images or sssd data. That data is often bound to a specific version of the software.
A simple example: Imagine that a database package is updated, and that update also includes a schema change which the post script will perform if a database is found. Now we run into the situation that the database is still running the *old* version in the currently active system and suddenly gets confronted with a changed and probably incompatible new data set belonging to the *new* version (with the new database daemon only available in the new snapshot which hasn't been booted yet). Due to the bind mount however the new data leaks into the current system. This is playing with fire.

Because of that I'm really interested to hear what the use case for adding /var/lib to the bind mounts would be.

**Comment by aplanas** [comment]
> I'll merge this for transactional-update 4.0 along with a man page for `tukit` itself (which is also still undocumented yet).

Sure. Will you prepare this?

> Just one note: I think we _never_ want to add _/var/lib_ to the list of bind mounts. The whole concept of transactional updates is based around the concept that an update can _not_ break the currently running system. By bind-mounting such a vital directory that whole concept is basically nullified - and on the contrary will cause significant side effects:

Indeed.

The issue is that there are some kubernetes installers that are doing this: updating `/var/lib`, adding directories there (like `/var/lib/rancher`).  This can be view as a bug in the installer (should use something different that `/var/lib`), but sadly is required as today.

> /var/lib contains a lot of data, including databases, docker images or sssd data. That data is often bound to a specific version of the software. A simple example: Imagine that a database package is updated, and that update also includes a schema change which the post script will perform if a database is found. Now we run into the situation that the database is still running the _old_ version in the currently active system and suddenly gets confronted with a changed and probably incompatible new data set belonging to the _new_ version (with the new database daemon only available in the new snapshot which hasn't been booted yet). Due to the bind mount however the new data leaks into the current system. This is playing with fire.

I will add as a comment in the documentation.

> Because of that I'm really interested to hear what the use case for adding /var/lib to the bind mounts would be.

Rancher.

**Comment by laenion** [comment]
> > Because of that I'm really interested to hear what the use case for adding /var/lib to the bind mounts would be.
> 
> Rancher.

But I guess to bind mount /var/lib/rancher would be enough then? Otherwise the whole concept of transactional-update is spoiled...

**Comment by aplanas** [comment]
> > > Because of that I'm really interested to hear what the use case for adding /var/lib to the bind mounts would be.
> > 
> > 
> > Rancher.
> 
> But I guess to bind mount /var/lib/rancher would be enough then? Otherwise the whole concept of transactional-update is spoiled...

IIUC not always, as this can be created by the installer.  This should be OK if:

1) The user create first the directory.

2) There is an specific partition created during the installation, and is already in fstab

**Comment by aplanas** [comment]
I changed the example, and I try to integrate the dangers of exposing /var/lib

File: man/tukit.conf.5.xml [hunk 1]
```diff
@@ -0,0 +1,179 @@
...
+            in a different partition, we should list both:
+            <programlisting>
+              BINDDIRS[0]="/opt"
+              BINDDIRS[1]="/usr/local"
+              BINDDIRS[2]="/usr/local/extra"
+              BINDDIRS[3]="/var/lib/images"
+            </programlisting>
+          </para>
+	  <para>
+	    Note that is some situations we should be tempted to
```
Comment (MaxiMii05) [inline man/tukit.conf.5.xml:153]: Just seeing this by chance and thought I could suggest fixing a typo and maybe a logical error (depending on how it's meant).
is -> in
should -> could

Comment (aplanas) [reply man/tukit.conf.5.xml:153]: I found and fixed the "is -> in", but I am not sure about the second one.

Thanks!

File: man/tukit.conf.5.xml [hunk 2]
```diff
@@ -0,0 +1,179 @@
...
+        </listitem>
+      </varlistentry>
+
+      <varlistentry>
+        <term><varname>BINDDIRS</varname></term>
+        <listitem>
+          <para>
+            Bind mount additional directories into the update
+            environment; to add multiple directories add a separate
+	    entry with a different id between the brackets for each.
```
Comment (laenion) [inline man/tukit.conf.5.xml:112]: The spare spaces from above could be inserted here at the beginning of the line ;-)

File: man/tukit.conf.5.xml [hunk 3]
```diff
@@ -0,0 +1,179 @@
...
+            rolled back) and the changes will be visible to the
+            running system immediately (breaking the transactional
+            concept and possibly resulting in a mix between old and
+            new files if there are also changes to the root file
+            system until the next reboot).
+          </para>
+          <para>
+            By default only /opt will be bind-mounted, as this
+            directory is often used by third-party packages, but the
+           applications in there are usually self-contained.
```
Comment (laenion) [inline man/tukit.conf.5.xml:127]: Missing space at beginning of line.

**Review by laenion — CHANGES_REQUESTED** [review]
Thanks a lot for the documentation! Generally I really like it, just a few nits inline.

File: man/tukit.conf.5.xml [hunk 4]
```diff
@@ -0,0 +1,179 @@
...
+
+  <refsect1>
+    <title>Description</title>
+
+    <para>This configuration file controls and defines the behaviour of
+    <citerefentry>
+      <refentrytitle>tukit</refentrytitle>
+      <manvolnum>8</manvolnum>
+    </citerefentry>.</para>
+    <para>The <refname>transactional-update.conf</refname> file uses INI
```
[the remark this answers is not in the document]
Comment (aplanas) [reply man/tukit.conf.5.xml]: Done!

File: man/tukit.conf.5.xml [hunk 5]
```diff
@@ -0,0 +1,179 @@
...
+            <literal>/sysroot</literal>.
+          </para>
+        </listitem>
+      </varlistentry>
+
+      <varlistentry>
+        <term><varname>LOCKFILE</varname></term>
+        <listitem>
+          <para>
+            Lock file path for tukit. By default is
```
[the remark this answers is not in the document]
Comment (aplanas) [reply man/tukit.conf.5.xml]: Done

File: man/tukit.conf.5.xml [hunk 6]
```diff
@@ -0,0 +1,179 @@
...
+          </para>
+        </listitem>
+      </varlistentry>
+
+      <varlistentry>
+        <term><varname>OVERLAY_DIR</varname></term>
+        <listitem>
+          <para>
+            Directory where the <literal>/etc</literal> overlays are
+            stored.  By default is <literal>/var/lib/overlay</literal>
```
[the remark this answers is not in the document]
Comment (aplanas) [reply man/tukit.conf.5.xml]: I reformatted all the `<para>`. I used a normal XML indentation here, expecting that spaces are not relevant.

File: man/tukit.conf.5.xml [hunk 7]
```diff
@@ -0,0 +1,179 @@
...
+        <listitem>
+          <para>
+            Bind mount additional directories into the update
+            environment; to add multiple directories add a separate
+	    entry with a different id between the brackets for each.
+          </para>
+          <para>
+            Be aware that transactions are only done the root file
+            system itself, i.e. any additional mounts added here will
+            *not* get a seperate snapshot (so their contents cannot be
```
[the remark this answers is not in the document]
Comment (aplanas) [reply man/tukit.conf.5.xml]: Done

**Pull request merged**